apps/emqx_rule_engine/test/emqx_rule_engine_SUITE.erl

```diff
@@ -2822,6 +2868,114 @@ t_get_rule_ids_by_action_reference_ingress_bridge(_Config) ->
     ),
     ok.
 
+%%------------------------------------------------------------------------------
+%% Test cases for rule metrics
+%%------------------------------------------------------------------------------
+
+-define(BRIDGE_TYPE, <<"mqtt">>).
+-define(BRIDGE_NAME, <<"bridge_over_troubled_water">>).
+-define(BRIDGE_CONFIG(QMODE), #{
+    <<"server">> => <<"127.0.0.1:1883">>,
+    <<"username">> => <<"user1">>,
+    <<"password">> => <<"">>,
+    <<"proto_ver">> => <<"v4">>,
+    <<"ssl">> => #{<<"enable">> => false},
+    <<"egress">> =>
+        #{
+            <<"local">> =>
+                #{
+                    <<"topic">> => <<"foo/#">>
+                },
+            <<"remote">> =>
+                #{
+                    <<"topic">> => <<"bar/${topic}">>,
+                    <<"payload">> => <<"${payload}">>,
+                    <<"qos">> => <<"${qos}">>,
+                    <<"retain">> => <<"${retain}">>
+                }
+        },
+    <<"resource_opts">> =>
+        #{
+            <<"health_check_interval">> => <<"5s">>,
+            <<"query_mode">> => QMODE,
+            <<"request_ttl">> => <<"3s">>,
+            <<"worker_pool_size">> => 1
+        }
+}).
+
+-define(SUCCESSS_METRICS, #{
+    matched := 1,
+    'actions.total' := 1,
+    'actions.failed' := 0,
+    'actions.success' := 1
+}).
+-define(FAIL_METRICS, #{
+    matched := 1,
+    'actions.total' := 1,
+    'actions.failed' := 1,
+    'actions.success' := 0
+}).
+
+t_rule_metrics_sync(_Config) ->
+    do_test_rule_metrics_success(<<"sync">>).
+
+t_rule_metrics_async(_Config) ->
+    do_test_rule_metrics_success(<<"async">>).
+
+t_rule_metrics_sync_fail(_Config) ->
+    do_test_rule_metrics_fail(<<"sync">>).
+
+t_rule_metrics_async_fail(_Config) ->
+    do_test_rule_metrics_fail(<<"async">>).
+
+do_test_rule_metrics_success(QMode) ->
+    ?assertMatch(
+        ?SUCCESSS_METRICS,
+        do_test_rule_metrics(QMode)
+    ).
+
+do_test_rule_metrics_fail(QMode) ->
+    ?assertMatch(
+        ?FAIL_METRICS,
+        do_test_rule_metrics(QMode)
+    ).
+
+do_test_rule_metrics(QMode) ->
+    BridgeId = create_bridge(?BRIDGE_TYPE, ?BRIDGE_NAME, ?BRIDGE_CONFIG(QMode)),
+    RuleId = <<"rule:test_metrics_bridge_action">>,
+    {ok, #{id := RuleId}} =
+        emqx_rule_engine:create_rule(
+            #{
+                id => RuleId,
+                sql => <<"SELECT * FROM \"topic/#\"">>,
+                actions => [BridgeId]
+            }
+        ),
+    timer:sleep(100),
```

Comment:
minor comment: to avoid the risk of creating flake test and not making the tests slow to run one could loop until the the expected condition is met and fail if it is not met after a certain amount of time (see https://github.com/klarna/snabbkaffe/blob/master/src/snabbkaffe.erl#L313)